proposed/2021/DotnetNugetTrustedSigners.md

```diff
@@ -0,0 +1,196 @@
+```
+
+#### `dotnet nuget trusted-signers trust-package`
+
+```
+Adds a trusted signer with the given name to the config, based on one or more packages.
+
+USAGE:
+    dotnet nuget trusted-signers trust-package [OPTIONS] <name> <package>...
+
+OPTIONS:
+        --allow-untrusted-root          Specifies if the certificate for the trusted signer should be allowed to chain to an untrusted root.
+        --author                        Specifies that the author signature of the package(s) should be trusted. This option is mutually exclusive with --repository.
+        --repository                    Specifies that the repository signature or countersignature of the package(s) should be trusted. This option is mutually exclusive with --author.
+        --owners                        Semi-colon separated list of trusted owners to further restrict the trust of a repository. Only valid when using the --repository option.
+    -h, --help                          Prints help information.
+    -v, --verbosity <LEVEL>             Set the MSBuild verbosity level. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic].
+        --interactive                   Allows command to stop and wait for user input or action.
+        --configfile                    The NuGet configuration file. If specified, only the settings from this file will be used. If not specified, the hierarchy of configuration files from the current directory will be used.
+```
+
+#### `dotnet nuget trusted-signers trust-certificate`
+
+```
+Adds a trusted signer with the given name to the config, based on a certificate fingerprint.
+
+Note: If a trusted signer with the given name already exists, the certificate item will be added to that signer. Otherwise a trusted author will be created with a certificate item from given certificate information.
+
+USAGE:
+    dotnet nuget trusted-signers trust-certificate [OPTIONS] <name> <fingerprint>
+
+OPTIONS:
+        --allow-untrusted-root          Specifies if the certificate for the trusted signer should be allowed to chain to an untrusted root.
+        --algorithm                     Specifies the hash algorithm used to calculate the certificate fingerprint. Defaults to SHA256. Values supported are SHA256, SHA384 and SHA512
+    -h, --help                          Prints help information.
+    -v, --verbosity <LEVEL>             Set the MSBuild verbosity level. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic].
+        --interactive                   Allows command to stop and wait for user input or action.
+        --configfile                    The NuGet configuration file. If specified, only the settings from this file will be used. If not specified, the hierarchy of configuration files from the current directory will be used.
+```
+
+#### `dotnet nuget trusted-signers trust-source`
```

Comment:
Let's say I've already trusted 15 package owners for my mid-sized project. But now I want to add one more package that is owned by someone I did not previously trust. What would be my flow to append one more trusted owner? Do I need to provide all 16 package owners to the `trust-source` command? What would my flow be if I have a large project with many more owners?

The reason I ask is that I recently tried to set up trusted signers for the [NuGetGallery codebase](https://github.com/nuget/nugetgallery). This proved to be difficult using `nuget.exe`, so I resorted to manually editing the `nuget.config` file instead.

Reply:
One idea may be to add a `trust-owner` command that appends one or more owners to a trusted repository. See @zivkan comment here: https://github.com/NuGet/Home/pull/10628/files/5263069456da2276d0f3143fd5bf616e933b7809#r589669891

Reply:
I have a similar question about a missing function in NuGet.exe:
After repo certificate rotation, we need to change the trustedSigners section to have two certs :
```
<trustedSigners>
  <repository name="nuget.org" serviceIndex="https://api.nuget.org/v3/index.json">
    <certificate fingerprint="0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D" hashAlgorithm="SHA256" allowUntrustedRoot="false" />
    <certificate fingerprint="5A2901D6ADA3D18260B9C6DFE2133C95D74B9EEF6AE0E5DC334C8454D1477DF4" hashAlgorithm="SHA256" allowUntrustedRoot="false" />
  </repository>
</trustedSigners>
```
But seems there is no command could do that. We have to edit config file manually.
Since we don't have that for NuGet.exe either, perhaps we can address that afterwards.